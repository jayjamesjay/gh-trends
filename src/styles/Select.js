import styled from 'styled-components';
import { light } from './Theme';

const Select = styled.select`
  max-width: 80%;
  margin: 0.5rem;
  padding: 0.5rem;
  background: ${props => props.theme.bgAdd};
  border: none;
  box-shadow: ${props => props.theme.shadow};
  font-size: 1.3rem;
  color: ${props => props.theme.color};
  text-align: center;
  text-transform: uppercase;
  appearance: none;
`;

Select.defaultProps = {
  theme: light
};

export default Select;
