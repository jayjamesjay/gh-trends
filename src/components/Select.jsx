import React from 'react';
import PropTypes from 'prop-types';
import StyledSelect from '../styles/Select';
import Option from '../styles/Option';
import { ImgIcon } from '../styles/Img';
import FilterImg from '../assets/img/filter-list.svg';

/**
 * @module Select
 */

/**
 * Language selection
 *
 * @property {string} curr - currently selected language
 * @property {function} onSelect - action fired when language is selected
 * @property {array} languages - list of languages to select from
 * @property {string} label - description of this language selection
 * @returns {SelectLang}
 */
export function SelectLang(props) {
  const { options, label } = props;
  const langs = ['All Languages'].concat(options);

  return (
    <>
      <ImgIcon src={FilterImg} alt={label} title={label} />
      <Select {...props} options={langs} label="Languages" />
    </>
  );
}

SelectLang.propTypes = {
  curr: PropTypes.string.isRequired,
  onSelect: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(PropTypes.string).isRequired,
  label: PropTypes.string.isRequired
};

/**
 * General selection
 *
 * @property {string} curr - currently selected option
 * @property {function} onSelect - action fired when option is selected
 * @property {array} options - list of options to select from
 * @property {string} label - description of this language selection
 * @returns {Select}
 */
export default function Select({ curr, onSelect, options, label }) {
  return (
    <StyledSelect value={curr} onChange={onSelect} aria-label={label}>
      {options.map(name => (
        <Option key={name} value={name}>
          {name}
        </Option>
      ))}
    </StyledSelect>
  );
}

Select.propTypes = {
  curr: PropTypes.string.isRequired,
  onSelect: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(PropTypes.string).isRequired,
  label: PropTypes.string.isRequired
};
