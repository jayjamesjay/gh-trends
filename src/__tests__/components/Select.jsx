import React from 'react';
import 'jest-styled-components';
import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import Select, { SelectLang } from '../../components/Select';

describe('<Select />', () => {
  const languages = ['JavaScript', 'CSS'];
  const label = 'Select language';
  const func = () => {};
  const curr = 'JavaScript';
  const select = () =>
    render(<Select curr={curr} onSelect={func} options={languages} label={label} />);

  it(`renders with default style`, () => {
    const component = renderer.create(
      <Select curr={curr} onSelect={func} options={languages} label={label} />,
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`has aria-label`, () => {
    select();
    expect(screen.queryByRole('combobox').getAttribute('aria-label')).toEqual('Select language');
  });

  it(`contains options`, () => {
    select();
    expect(screen.queryAllByRole('option')).toHaveLength(2);
  });
});

describe('<SelectLang />', () => {
  const languages = ['JavaScript', 'CSS'];
  const label = 'Select language';
  const func = () => {};
  const curr = 'JavaScript';
  const select = () =>
    render(<SelectLang curr={curr} onSelect={func} options={languages} label={label} />);

  it(`renders with default style`, () => {
    const component = renderer.create(
      <SelectLang curr={curr} onSelect={func} options={languages} label={label} />,
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`has aria-label`, () => {
    select();
    expect(screen.queryByRole('combobox').getAttribute('aria-label')).toEqual('Languages');
  });

  it(`contains image`, () => {
    select();
    expect(screen.queryByRole('img').getAttribute('alt')).toEqual(label);
  });
});
