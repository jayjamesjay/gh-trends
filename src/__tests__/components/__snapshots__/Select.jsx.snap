// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> renders with default style 1`] = `
<DocumentFragment>
  .c0 {
  max-width: 80%;
  margin: 0.5rem;
  padding: 0.5rem;
  background: #fff;
  border: none;
  border-radius: 0.5rem;
  box-shadow: 0px 3px 5px 1px rgba(217,217,217,0.82);
  font-family: 'Noto Sans HK',sans-serif;
  font-size: 1.3rem;
  color: #000;
  text-align: center;
  text-transform: uppercase;
  appearance: none;
}

.c1 {
  font-size: 1.2rem;
  line-break: loose;
  text-align: center;
  text-transform: uppercase;
}

<select
    aria-label="Select language"
    class="c0"
  >
    <option
      class="c1"
      value="JavaScript"
    >
      JavaScript
    </option>
    <option
      class="c1"
      value="CSS"
    >
      CSS
    </option>
  </select>
</DocumentFragment>
`;

exports[`<SelectLang /> renders with default style 1`] = `
<DocumentFragment>
  .c0 {
  width: 100%;
  height: auto;
  filter: invert(0%);
}

.c1 {
  margin-top: 0.5rem;
  width: auto;
  height: 3rem;
}

<img
    alt="Select language"
    class="c0 c1"
    src="test-file-stub"
    title="Select language"
  />
  .c0 {
  max-width: 80%;
  margin: 0.5rem;
  padding: 0.5rem;
  background: #fff;
  border: none;
  border-radius: 0.5rem;
  box-shadow: 0px 3px 5px 1px rgba(217,217,217,0.82);
  font-family: 'Noto Sans HK',sans-serif;
  font-size: 1.3rem;
  color: #000;
  text-align: center;
  text-transform: uppercase;
  appearance: none;
}

.c1 {
  font-size: 1.2rem;
  line-break: loose;
  text-align: center;
  text-transform: uppercase;
}

<select
    aria-label="Languages"
    class="c0"
  >
    <option
      class="c1"
      value="All Languages"
    >
      All Languages
    </option>
    <option
      class="c1"
      value="JavaScript"
    >
      JavaScript
    </option>
    <option
      class="c1"
      value="CSS"
    >
      CSS
    </option>
  </select>
</DocumentFragment>
`;
